dandi/files/zarr.py

```diff
@@ -540,14 +541,38 @@ def mkzarr() -> str:
 def _upload_zarr_file(
     storage_session: RESTFullAPIClient, upload_url: str, item: UploadItem
 ) -> int:
-    with item.filepath.open("rb") as fp:
-        storage_session.put(
-            upload_url,
-            data=fp,
-            json_resp=False,
-            retry_if=_retry_zarr_file,
-            headers={"Content-MD5": item.base64_digest},
-        )
+    with ExitStack() as stack:
+        if os.environ.get("DANDI_DEVEL_INSTRUMENT_REQUESTS_SUPERLEN"):
+            from requests.utils import super_len
+
+            def new_super_len(o):
+                try:
+                    n = super_len(o)
+                except Exception:
+                    lgr.debug(
+                        "requests.utils.super_len() failed while uploading %s:",
+                        item.filepath,
+                        exc_info=True,
+                    )
+                    raise
+                else:
+                    lgr.debug(
+                        "requests.utils.super_len() reported %d while uploading %s",
+                        n,
+                        item.filepath,
+                    )
+                    return n
+
+            stack.enter_context(patch("requests.models.super_len", new_super_len))
+
+        with item.filepath.open("rb") as fp:
+            storage_session.put(
+                upload_url,
+                data=fp,
+                json_resp=False,
+                retry_if=_retry_zarr_file,
+                headers={"Content-MD5": item.base64_digest},
+            )
```

Comment:
nice design pattern on many aspects!  

@asmacdo @candleindark - please review/take into your toolbelt on how to 
- add optional context managers to the piece of code without duplication
- instrument/patch some functionality (outside or not) temporarily